pkg/apis/agones/v1/gameserver.go

```diff
@@ -926,3 +926,83 @@ func (gs *GameServer) Patch(delta *GameServer) ([]byte, error) {
 	result, err = json.Marshal(patch)
 	return result, errors.Wrapf(err, "error creating json for patch for GameServer %s", gs.ObjectMeta.Name)
 }
+
+// UpdateCount increments or decrements a CounterStatus on a Game Server by the given amount.
+// TODO: Should we log on noncomplemention or return false? Issue #2716 asks for "silent no-op"
+func (gs *GameServer) UpdateCount(name string, action string, amount int64) {
+	if !(action == GameServerAllocationIncrement || action == GameServerAllocationDecrement) || amount < 0 {
+		return
+	}
+	if counter, ok := gs.Status.Counters[name]; ok {
+		cnt := counter.Count
+		if action == GameServerAllocationIncrement {
+			cnt += amount
+		}
+		if action == GameServerAllocationDecrement {
+			cnt -= amount
+		}
+		// TODO: Can we force the Capacity to update first before updating the Count if CounterAction updates both?
+		if (cnt < 0) || (cnt > counter.Capacity) {
+			return
+		}
+		counter.Count = cnt
+		gs.Status.Counters[name] = counter
+	}
+}
+
+// UpdateCounterCapacity updates the CounterStatus Capacity to the given capacity. Silent no-op.
+// TODO: Do anything if Capacity is updated to be less than the current Count?
+func (gs *GameServer) UpdateCounterCapacity(name string, capacity int64) {
+	if capacity < 0 {
+		return
+	}
+	if counter, ok := gs.Status.Counters[name]; ok {
+		counter.Capacity = capacity
+		gs.Status.Counters[name] = counter
+	}
+}
+
+// UpdateListCapacity updates the ListStatus Capacity to the given capacity. Silent no-op.
+func (gs *GameServer) UpdateListCapacity(name string, capacity int64) {
+	if capacity < 0 || capacity > 1000 {
+		return
+	}
+	if list, ok := gs.Status.Lists[name]; ok {
+		list.Capacity = capacity
+		gs.Status.Lists[name] = list
+	}
+}
+
+// AppendListValues adds given the values to the ListStatus Values list. Any duplicates are silently
+// dropped. Silent no-op.
+func (gs *GameServer) AppendListValues(name string, values []string) {
+	if len(values) == 0 {
+		return
+	}
+	if list, ok := gs.Status.Lists[name]; ok {
+		vals := []string{}
+		// Maintain ordering so new values are appended to the end of list.Values
+		vals = append(vals, list.Values...)
+		vals = append(vals, values...)
+		vals = removeDuplicates(vals)
```

Comment:
Wondering if there is a SortedSet implementation we might be able to use - make our lives easier?

Reply:
There's https://github.com/wangjia184/sortedset which unless we changed list.Values to a sortedset wouldn't give us any time complexity gains here since we'd need to convert from []string to sortedset and then back to []string.

Reply:
Sounds like too much work 👍🏻